File: ai-service/app/security/csrf.py
"""CSRF (Cross-Site Request Forgery) protection for LAYA AI Service.

This module provides CSRF token generation and validation to protect against
CSRF attacks on state-changing operations (POST, PUT, DELETE, PATCH).

CSRF tokens are signed using the application's JWT secret key to ensure
authenticity without requiring server-side state storage.
"""

import secrets
from datetime import datetime, timedelta, timezone
from typing import Callable, List, Optional

from fastapi import HTTPException, Request, Response, status
from jose import JWTError, jwt
from starlette.middleware.base import BaseHTTPMiddleware

from app.config import settings


def generate_csrf_token(duration_minutes: Optional[int] = None) -> str:
    """Generate a cryptographically secure CSRF token.

    The token is a JWT containing:
    - Random nonce for uniqueness
    - Expiration timestamp
    - Token type identifier

    Args:
        duration_minutes: Token validity duration in minutes
                         (default: from settings.csrf_token_expire_minutes)

    Returns:
        str: Signed CSRF token (JWT)
    """
    # Use configured expiration time if not specified
    if duration_minutes is None:
        duration_minutes = settings.csrf_token_expire_minutes

    # Generate a random nonce for uniqueness
    nonce = secrets.token_urlsafe(32)

    # Calculate expiration time
    expires_at = datetime.now(timezone.utc) + timedelta(minutes=duration_minutes)

    # Create token payload
    payload = {
        "nonce": nonce,
        "exp": expires_at.timestamp(),
        "type": "csrf",
    }

    # Sign token using JWT secret
    token = jwt.encode(payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm)

    return token


def validate_csrf_token(token: str) -> bool:
    """Validate a CSRF token.

    Verifies:
    - Token signature is valid
    - Token has not expired
    - Token type is 'csrf'

    Args:
        token: The CSRF token to validate

    Returns:
        bool: True if token is valid, False otherwise
    """
    try:
        # Decode and verify token
        payload = jwt.decode(
            token,
            settings.jwt_secret_key,
            algorithms=[settings.jwt_algorithm],
        )

        # Verify token type
        if payload.get("type") != "csrf":
            return False

        # JWT library automatically checks expiration
        return True

    except JWTError:
        # Token is invalid (expired, malformed, wrong signature, etc.)
        return False


def get_csrf_exempt_paths() -> List[str]:
    """Get list of paths exempt from CSRF protection.

    These paths are exempt because they:
    - Are read-only (GET, HEAD, OPTIONS)
    - Are used by external webhooks (can't provide CSRF tokens)
    - Are health checks or monitoring endpoints

    Returns:
        List[str]: List of path prefixes exempt from CSRF protection
    """
    return [
        "/",  # Health check
        "/docs",  # API documentation
        "/openapi.json",  # OpenAPI spec
        "/api/v1/webhook",  # Webhook endpoints (external calls)
    ]


class CSRFProtectionMiddleware(BaseHTTPMiddleware):
    """CSRF protection middleware for FastAPI application.

    This middleware validates CSRF tokens on all state-changing requests
    (POST, PUT, DELETE, PATCH) except for exempt paths.

    CSRF tokens must be provided in the X-CSRF-Token header.
    """

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """Process request and validate CSRF token if needed.

        Args:
            request: The incoming request
            call_next: The next middleware/handler in the chain

        Returns:
            Response: The response from the handler

        Raises:
            HTTPException: 403 if CSRF token is missing or invalid
        """
        # Get HTTP method
        method = request.method.upper()

        # Only check CSRF for state-changing methods
        state_changing_methods = ["POST", "PUT", "DELETE", "PATCH"]

        if method in state_changing_methods:
            # Check if path is exempt
            path = request.url.path
            exempt_paths = get_csrf_exempt_paths()

            is_exempt = any(path.startswith(exempt_path) for exempt_path in exempt_paths)

            if not is_exempt:
                # Get CSRF token from header
                csrf_token = request.headers.get("X-CSRF-Token")

                if not csrf_token:
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="CSRF token missing. Include X-CSRF-Token header.",
                    )

                # Validate token
                if not validate_csrf_token(csrf_token):
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="CSRF token invalid or expired.",
                    )

        # Process the request
        response = await call_next(request)

        return response


def get_csrf_protection_middleware() -> Callable:
    """Get CSRF protection middleware for the FastAPI application.

    This function is kept for backward compatibility but returns the class.

    Returns:
        type: CSRFProtectionMiddleware class
    """
    return CSRFProtectionMiddleware


def get_csrf_token_from_request(request: Request) -> Optional[str]:
    """Extract CSRF token from request headers.

    Args:
        request: The incoming request

    Returns:
        Optional[str]: CSRF token if present, None otherwise
    """
    return request.headers.get("X-CSRF-Token")

File: parent-portal/lib/security/csrf.ts
/**
 * CSRF (Cross-Site Request Forgery) protection for LAYA Parent Portal.
 *
 * This module provides CSRF token management for protecting against
 * CSRF attacks on state-changing operations (POST, PUT, DELETE, PATCH).
 *
 * CSRF tokens are fetched from the backend and included in request headers.
 * Tokens are JWT-based and validated server-side for authenticity.
 */

import { jwtVerify, decodeJwt } from 'jose';

/**
 * CSRF token storage interface
 */
interface CSRFTokenStore {
  token: string | null;
  expiresAt: number | null;
}

// In-memory token storage (tokens are also stored in HttpOnly cookies on backend)
const tokenStore: CSRFTokenStore = {
  token: null,
  expiresAt: null,
};

/**
 * Configuration for CSRF token management
 */
export interface CSRFConfig {
  /**
   * Backend API URL for fetching CSRF tokens
   */
  apiUrl: string;
  /**
   * Token expiration buffer in minutes (refresh token this many minutes before expiry)
   */
  expirationBufferMinutes?: number;
}

/**
 * Default configuration
 */
const DEFAULT_CONFIG: Required<CSRFConfig> = {
  apiUrl: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000',
  expirationBufferMinutes: 5,
};

let config: Required<CSRFConfig> = DEFAULT_CONFIG;

/**
 * Initialize CSRF protection with custom configuration
 *
 * @param userConfig - Custom configuration options
 */
export function initCSRFProtection(userConfig: Partial<CSRFConfig>): void {
  config = {
    ...DEFAULT_CONFIG,
    ...userConfig,
  };
}

/**
 * Fetch a new CSRF token from the backend
 *
 * Makes a GET request to the /api/v1/csrf-token endpoint.
 * The backend also sets an HttpOnly cookie for additional security.
 *
 * @returns Promise resolving to the CSRF token
 * @throws Error if token fetch fails
 */
export async function fetchCSRFToken(): Promise<string> {
  try {
    const response = await fetch(`${config.apiUrl}/api/v1/csrf-token`, {
      method: 'GET',
      credentials: 'include', // Include cookies for HttpOnly cookie
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error(
        `Failed to fetch CSRF token: ${response.status} ${response.statusText}`
      );
    }

    const data = await response.json();

    if (!data.csrf_token) {
      throw new Error('CSRF token not found in response');
    }

    // Store token and expiration
    const token = data.csrf_token as string;
    storeCSRFToken(token);

    return token;
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`CSRF token fetch error: ${error.message}`);
    }
    throw new Error('CSRF token fetch error: Unknown error');
  }
}

/**
 * Store a CSRF token and calculate its expiration
 *
 * Decodes the JWT token to extract the expiration timestamp.
 *
 * @param token - The CSRF token to store
 */
export function storeCSRFToken(token: string): void {
  try {
    // Decode JWT to get expiration (without verification - backend handles that)
    const payload = decodeJwt(token);

    // Extract expiration timestamp
    const expiresAt = payload.exp ? payload.exp * 1000 : null; // Convert to milliseconds

    tokenStore.token = token;
    tokenStore.expiresAt = expiresAt;
  } catch (error) {
    // If decoding fails, store token without expiration info
    tokenStore.token = token;
    tokenStore.expiresAt = null;
  }
}

/**
 * Get the current CSRF token from storage
 *
 * @returns The current CSRF token or null if not set
 */
export function getCSRFToken(): string | null {
  return tokenStore.token;
}

/**
 * Check if the current CSRF token is expired or about to expire
 *
 * Returns true if:
 * - No token is stored
 * - Token is expired
 * - Token will expire within the configured buffer time
 *
 * @returns True if token needs refresh, false otherwise
 */
export function isCSRFTokenExpired(): boolean {
  if (!tokenStore.token || !tokenStore.expiresAt) {
    return true;
  }

  const now = Date.now();
  const bufferMs = config.expirationBufferMinutes * 60 * 1000;
  const expiresWithBuffer = tokenStore.expiresAt - bufferMs;

  return now >= expiresWithBuffer;
}

/**
 * Get a valid CSRF token, fetching a new one if necessary
 *
 * This is the main function to use when you need a CSRF token for a request.
 * It automatically handles token refresh if the current token is expired.
 *
 * @returns Promise resolving to a valid CSRF token
 * @throws Error if token fetch fails
 */
export async function getValidCSRFToken(): Promise<string> {
  if (isCSRFTokenExpired()) {
    return await fetchCSRFToken();
  }

  return tokenStore.token as string;
}

/**
 * Clear the stored CSRF token
 *
 * Call this on logout or when you want to force a token refresh.
 */
export function clearCSRFToken(): void {
  tokenStore.token = null;
  tokenStore.expiresAt = null;
}

/**
 * Get list of HTTP methods that require CSRF protection
 *
 * These are state-changing methods that should include a CSRF token.
 *
 * @returns Array of HTTP methods requiring CSRF protection
 */
export function getCSRFProtectedMethods(): string[] {
  return ['POST', 'PUT', 'DELETE', 'PATCH'];
}

/**
 * Get list of paths exempt from CSRF protection
 *
 * These paths are exempt because they:
 * - Are read-only (GET, HEAD, OPTIONS)
 * - Are used by external webhooks (can't provide CSRF tokens)
 * - Are health checks or monitoring endpoints
 *
 * @returns Array of path prefixes exempt from CSRF protection
 */
export function getCSRFExemptPaths(): string[] {
  return [
    '/api/health',
    '/api/docs',
    '/api/openapi.json',
    '/api/v1/webhook',
  ];
}

/**
 * Check if a given path requires CSRF protection
 *
 * @param path - The URL path to check
 * @param method - The HTTP method (defaults to 'POST')
 * @returns True if path requires CSRF protection, false otherwise
 */
export function requiresCSRFProtection(
  path: string,
  method: string = 'POST'
): boolean {
  // Check if method requires CSRF protection
  const protectedMethods = getCSRFProtectedMethods();
  if (!protectedMethods.includes(method.toUpperCase())) {
    return false;
  }

  // Root path is exempt (health check)
  if (path === '/') {
    return false;
  }

  // Check if path is exempt (using exact match or prefix match)
  const exemptPaths = getCSRFExemptPaths();
  const isExempt = exemptPaths.some((exemptPath) =>
    path === exemptPath || path.startsWith(exemptPath + '/')
  );

  return !isExempt;
}

/**
 * Validate a CSRF token structure (client-side check only)
 *
 * This performs a basic validation by decoding the JWT and checking:
 * - Token signature structure (not cryptographic verification)
 * - Token has not expired
 * - Token type is 'csrf'
 *
 * Note: Full validation happens server-side. This is just for client-side
 * checks to avoid sending obviously invalid tokens.
 *
 * @param token - The CSRF token to validate
 * @returns True if token appears valid, false otherwise
 */
export function validateCSRFTokenStructure(token: string): boolean {
  try {
    const payload = decodeJwt(token);

    // Check token type
    if (payload.type !== 'csrf') {
      return false;
    }

    // Check expiration
    if (!payload.exp) {
      return false;
    }

    const now = Math.floor(Date.now() / 1000);
    if (payload.exp < now) {
      return false;
    }

    return true;
  } catch {
    // Token is malformed
    return false;
  }
}

/**
 * Create request headers with CSRF token included
 *
 * Use this helper to automatically include the CSRF token in your requests.
 *
 * @param existingHeaders - Existing headers to merge with
 * @returns Promise resolving to headers object with CSRF token included
 * @throws Error if token fetch fails
 */
export async function createCSRFHeaders(
  existingHeaders: HeadersInit = {}
): Promise<Headers> {
  const token = await getValidCSRFToken();
  const headers = new Headers(existingHeaders);

  headers.set('X-CSRF-Token', token);

  return headers;
}

/**
 * Fetch wrapper that automatically includes CSRF token for state-changing requests
 *
 * Use this instead of the standard fetch() for API calls that require CSRF protection.
 *
 * @param url - The URL to fetch
 * @param options - Fetch options (method, headers, body, etc.)
 * @returns Promise resolving to the Response
 */
export async function fetchWithCSRF(
  url: string,
  options: RequestInit = {}
): Promise<Response> {
  const method = options.method?.toUpperCase() || 'GET';

  // Extract path from URL for checking
  const urlObj = new URL(url, config.apiUrl);
  const path = urlObj.pathname;

  // Only add CSRF token for protected methods and paths
  if (requiresCSRFProtection(path, method)) {
    const headers = await createCSRFHeaders(options.headers);

    return fetch(url, {
      ...options,
      headers,
      credentials: 'include', // Include cookies
    });
  }

  // For non-protected requests, just pass through
  return fetch(url, {
    ...options,
    credentials: 'include', // Still include cookies for auth
  });
}

/**
 * React hook for accessing CSRF token utilities
 *
 * @returns Object with CSRF token utilities
 */
export function useCSRF() {
  return {
    fetchToken: fetchCSRFToken,
    getToken: getCSRFToken,
    getValidToken: getValidCSRFToken,
    clearToken: clearCSRFToken,
    isExpired: isCSRFTokenExpired,
    validateStructure: validateCSRFTokenStructure,
    createHeaders: createCSRFHeaders,
    fetchWithCSRF,
  };
}

/**
 * Export token store for testing purposes only
 * @internal
 */
export const __internal = {
  tokenStore,
  config: () => config,
  resetConfig: () => {
    config = DEFAULT_CONFIG;
  },
};
